src/services/compositionRootService.ts

```diff
@@ -1,113 +1,144 @@
+import { SettingsManager } from '../settings/settingsManager';
+import { VSCodeManager } from '../vscode/vscodeManager';
+import { IntegrityManager } from '../integrity/integrityManager';
+
+type Class = new (...args: any[]) => any;
 
 export class CompositionRootService {
-  private readonly container: any[];
+  private readonly container: Container;
+  private injectableClasses: ReadonlyArray<[Class, ServiceIdentifierOrFunc[]]>;
+
+  constructor(private context: models.IVSCodeExtensionContext) {
+    this.container = new Container({ defaultScope: 'Singleton' });
+    this.init();
+    this.initDecorations();
+    this.initBindings();
+  }
+
+  public get<T>(serviceIdentifier: interfaces.ServiceIdentifier<T>): T {
+    return this.container.get<T>(serviceIdentifier);
+  }
 
-  constructor(context: IVSCodeExtensionContext) {
-    this.container = [];
-    if (context) {
-      this.register(context);
-    }
+  public dispose() {
+    this.injectableClasses
+      .map(injectableClass => injectableClass[0])
+      .forEach(klass => {
+        Reflect.deleteMetadata(METADATA_KEY.PARAM_TYPES, klass);
+        Reflect.deleteMetadata(METADATA_KEY.TAGGED, klass);
+      });
   }
 
-  public get<T>(identifier: ServiceIdentifier): T {
-    const result = this.container.find(element => {
-      return element.identifier === identifier;
+  private init(): void {
+    // Each tuple has the form of (class, array of parameters)
+    // The array of parameters MUST be declared in the precise order
+    // they are declared in the constructor of each class
+    this.injectableClasses = [
+      [
+        ExtensionManager,
+        [
+          models.SYMBOLS.IVSCodeManager,
+          models.SYMBOLS.IConfigManager,
+          models.SYMBOLS.ISettingsManager,
+          models.SYMBOLS.INotificationManager,
+          models.SYMBOLS.IIconsGenerator,
+          models.SYMBOLS.IProjectAutoDetectionManager,
+          models.SYMBOLS.IIntegrityManager,
+        ],
+      ],
+      [ConfigManager, [models.SYMBOLS.IVSCodeManager]],
+      [LanguageResourceManager, [models.SYMBOLS.ILocale]],
+      [
+        IconsGenerator,
+        [models.SYMBOLS.IVSCodeManager, models.SYMBOLS.IConfigManager],
+      ],
+      [
+        NotificationManager,
+        [
+          models.SYMBOLS.IVSCodeManager,
+          models.SYMBOLS.ILanguageResourceManager,
+        ],
+      ],
+      [
+        ProjectAutoDetectionManager,
+        [models.SYMBOLS.IVSCodeManager, models.SYMBOLS.IConfigManager],
+      ],
+      [SettingsManager, [models.SYMBOLS.IVSCodeManager]],
+      [
+        VSCodeManager,
+        [models.SYMBOLS.IVSCode, models.SYMBOLS.IVSCodeExtensionContext],
+      ],
+      [IntegrityManager, []],
+    ];
+  }
+
+  private initDecorations(): void {
+    this.injectableClasses.forEach(injectableClass => {
+      // declare classes as injectables
+      const klass: Class = injectableClass[0];
+      decorate(injectable(), klass);
```

Comment:
This line is giving a test error:
`Cannot apply @Injectable decorator multiple times.`

Please refer to [this issue](https://github.com/inversify/InversifyJS/issues/563) ànd the author for a possible workaround.

Reply:
Is this happening when you use the `-w` switch?

Reply:
Nope, it is not giving any test error if using the `-w` flag.

Reply:
You need to be more specific. As you can see the tests are passing in all ci suites and on my local machines.

Reply:
I was running the test using my Windows 10 OS, five test cases failed if running the `test` command without the `-w` flag. Looks like the related issue wasn't related at all, since the `watch` mode is giving me all test passing.

Reply:
Just test it on a `Windows 10 1809`. Everything checks out. Also, test it on `MacOS High Sierra` and `Ubuntu 18.04` and all tests pass.

You need to troubleshoot your installation. I suggest to do a clean `npm i` and then `npm t`.

Reply:
Can anyone else confirm @KingDarBoja issue?

Reply:
I accidentally stumbled on the issue. The error is presented only when you use `Launch Tests` that uses the `vscode` test runner.